consensus/tests/checkpoint.rs

```diff
@@ -43,48 +50,201 @@ mod tests {
+
+        // This invariant needs to be held for the position of `a`
+        assert!(max_epoch_slot > a_curr_epoch_slot.0 + 2);
+
+        // Assume there is a next block in the slot directly preceeding the block for `a`
+        let added_slots = thread_rng().gen_range(a_curr_epoch_slot.0 + 2..max_epoch_slot);
+
+        let added_blocks = gen_num_blocks_in_slots(
+            DEFAULT_SLOT_FILL_RATE,
+            DEFAULT_SLOT_FILL_RATE_DELTA,
+            added_slots as f64,
+        );
+
+        b.body.consensus_state.curr_global_slot.slot_number.0 =
+            a_curr_epoch_slot.0 + added_slots + 1;
+        b.body.consensus_state.blockchain_length.0 = a_curr_epoch_length + added_blocks as u32 + 1;
+    }
+
     #[test]
     #[wasm_bindgen_test]
-    fn test_is_short_range() {
+    fn equal_state_in_short_fork_range() {
         let mut genesis: ProtocolState = Default::default();
         init_checkpoints(&mut genesis).unwrap();
-        let mut c0: ProtocolStateChain = ProtocolStateChain(vec![]);
-        let mut c1: ProtocolStateChain = ProtocolStateChain(vec![]);
-        let mut c3: ProtocolStateChain = ProtocolStateChain(vec![]);
-        let mut b0: ProtocolState = Default::default();
-        b0.body.consensus_state.blockchain_length = Length(0);
-        b0.body.consensus_state.curr_global_slot = GlobalSlot {
-            slot_number: GlobalSlotNumber(0),
-            slots_per_epoch: Length(7140),
-        };
-        let mut b1: ProtocolState = Default::default();
-        b1.body.consensus_state.blockchain_length = Length(1);
-        b1.body.consensus_state.curr_global_slot = GlobalSlot {
-            slot_number: GlobalSlotNumber(1),
-            slots_per_epoch: Length(7140),
-        };
 
-        c0.push(b0).unwrap();
-        c1.push(b1).unwrap();
+        let mut a: ProtocolState = Default::default();
+        let mut b: ProtocolState = Default::default();
+
+        gen_spot(&mut a);
+        gen_spot(&mut b);
+
+        let c0: ProtocolStateChain = ProtocolStateChain(vec![a]);
+        let c1: ProtocolStateChain = ProtocolStateChain(vec![b]);
+
+        assert_eq!(is_short_range(&c0, &c1).unwrap(), true);
+    }
+
+    #[test]
+    #[wasm_bindgen_test]
+    fn gen_spot_pair_short_aligned_generates_pairs_of_states_in_short_fork_range() {
+        // Both states will share their staking epoch checkpoints.
+        let mut genesis: ProtocolState = Default::default();
+        init_checkpoints(&mut genesis).unwrap();
+
+        let mut a: ProtocolState = Default::default();
+        let mut b: ProtocolState = Default::default();
+
+        gen_spot(&mut a);
+        gen_spot(&mut b);
+        gen_spot_pair_common_checkpoints(&mut a, &mut b, 0);
+
+        let c0: ProtocolStateChain = ProtocolStateChain(vec![a]);
+        let c1: ProtocolStateChain = ProtocolStateChain(vec![b]);
+
         assert_eq!(is_short_range(&c0, &c1).unwrap(), true);
-        assert_eq!(is_short_range(&c1, &c0).unwrap(), true);
```

Comment:
This assert seems commented

Reply:
> Bug: While going through the is_short_range function from the spec:
> 
> ```rust
> fn isShortRange(C1, C2) -> bool
> {
>     // Get consensus state from top blocks of each chain
>     let S1 = cState(C1);
>     let S2 = cState(C2);
> 
>     let check = | S1, S2 | {
>       if S1.epoch_count == S2.epoch_count + 1  && epochSlot(S2) >= 2/3*slots_per_epoch {
>         // S1 is one epoch ahead of S2 and S2 is not in the seed update range
>         return S1.staking_epoch_data.lock_checkpoint == S2.next_epoch_data.lock_checkpoint
>       }
>       else {
>         return false
>       }
>     };
> 
>     if S1.epoch_count == S2.epoch_count {
>         // Simple case: blocks have same previous epoch, so compare previous epochs' lock_checkpoints
>         return S1.staking_epoch_data.lock_checkpoint == S2.staking_epoch_data.lock_checkpoint
>     }
>     else {
>         // Check for previous epoch case using both orientations
>         return check(S1, S2) || check(S2, S1)
>     }
> }
> ```
> 
> it appears the impl here is missing the both way check: `return check(S1, S2) || check(S2, S1)`
> 
> we should be checking it both ways


in the previous conversation, we resolved both ways check at is_short_range()  such that we don't need to check both ways again in the test.